specs/experimental/cannon-fault-proof-vm-mt.md

```diff
@@ -0,0 +1,511 @@
+
+Writing to and reading from standard output, input and error streams have no effect on the FPVM state.
+FPVM implementations may use them for debugging purposes as long as I/O is stateless.
+
+All I/O operations are restricted to a maximum of 4 bytes per operation.
+Any read or write syscall request exceeding this limit will be truncated to 4 bytes.
+Consequently, the return value of read/write syscalls is at most 4, indicating the actual number of bytes read/written.
+
+### Standard Streams
+
+Writing to stderr/stdout standard stream always succeeds with the write count input returned,
+effectively continuing execution without writing work.
+Reading from stdin has no effect other than to return zero and errno set to 0, signalling that there is no input.
+
+### Hint Communication
+
+Hint requests and responses have no effect on the VM state other than setting the `$v0` return
+register to the requested read/write count.
+VM implementations may utilize hints to setup subsequent pre-image requests.
+
+### Pre-image Communication
+
+The `preimageKey` and `preimageOffset` state are updated via read/write syscalls to the pre-image
+read and write file descriptors (see [I/O](#io)).
+The `preimageKey` buffers the stream of bytes written to the pre-image write fd.
+The `preimageKey` buffer is shifted to accommodate new bytes written to the end of it.
+A write also resets the `preimageOffset` to 0, indicating the intent to read a new pre-image.
+
+When handling pre-image reads, the `preimageKey` is used to lookup the pre-image data from an Oracle.
+A max 4-byte chunk of the pre-image at the `preimageOffset` is read to the specified address.
+Each read operation increases the `preimageOffset` by the number of bytes requested
+(truncated to 4 bytes and subject to alignment constraints).
+
+#### Pre-image I/O Alignment
+
+As mentioned earlier in [memory](#memory), all memory operations are 4-byte aligned.
+Since pre-image I/O occurs on memory, all pre-image I/O operations must strictly adhere to alignment boundaries.
+This means the start and end of a read/write operation must fall within the same alignment boundary.
+If an operation were to violate this, the input `count` of the read/write syscall must be
+truncated such that the effective address of the last byte read/written matches the input effective address.
+
+The VM must read/write the maximum amount of bytes possible without crossing the input address alignment boundary.
+For example, the effect of a write request for a 3-byte aligned buffer must be exactly 3 bytes.
+If the buffer is misaligned, then the VM may write less than 3 bytes depending on the size of the misalignment.
+
+## Exceptions
+
+The FPVM may raise an exception rather than output a post-state to signal an invalid state
+transition. Nominally, the FPVM must raise an exception in at least the following cases:
+
+- Invalid instruction (either via an invalid opcode or an instruction referencing registers
```

Comment:
Let's expand this list to include the invalid thread states.

Reply:
Updated by adding:
```
- Invalid thread state:
  - There are no threads - both thread stacks are empty.
  - The active thread stack is empty.
```

The first bullet is kind of redundant given the second bullet, but we do explicitly check for it in MIPS2.sol.

Also wonder if we should make that [MIPS2.sol check](https://github.com/ethereum-optimism/optimism/blob/3210a8c6bb7f98639213fb49718919d84aa8774a/packages/contracts-bedrock/src/cannon/MIPS2.sol#L165-L167) stricter by only verifying that the active thread stack is not empty?

Reply:
Yup, we should add that check. Created an [issue](https://github.com/ethereum-optimism/optimism/issues/12160) for it.